modules/sim_2d/sim_2d_component.cc

```diff
@@ -100,93 +96,69 @@ void SimControl::InitTimerAndIO() {
       node_->CreateWriter<PredictionObstacles>(FLAGS_prediction_topic);
 
   // Start timer to publish localization and chassis messages.
-  sim_control_timer_.reset(new cyber::Timer(
-      kSimControlIntervalMs, [this]() { this->RunOnce(); }, false));
-  sim_prediction_timer_.reset(
-      new cyber::Timer(kSimPredictionIntervalMs,
-                       [this]() { this->PublishDummyPrediction(); }, false));
+  sim_2d_timer_.reset(new cyber::Timer(
+      kSim2DComponentIntervalMs, [this]() { this->SimStep(); }, false));
+  sim_prediction_timer_.reset(new cyber::Timer(
+      kSimPredictionIntervalMs, [this]() { this->PublishDummyPrediction(); },
+      false));
 }
 
-void SimControl::Init(bool set_start_point, double start_velocity,
-                      double start_acceleration) {
+void Sim2DComponent::InitState(bool set_start_point, double start_velocity,
+                               double start_acceleration) {
   if (set_start_point && !FLAGS_use_navigation_mode) {
     InitStartPoint(start_velocity, start_acceleration);
   }
+  InitTimerAndIO();
 }
 
-void SimControl::InitStartPoint(double start_velocity,
-                                double start_acceleration) {
+void Sim2DComponent::InitStartPoint(double start_velocity,
+                                    double start_acceleration) {
   TrajectoryPoint point;
-  // Use the latest localization position as start point,
   // fall back to a dummy point from map
-  localization_reader_->Observe();
-  if (localization_reader_->Empty()) {
-    start_point_from_localization_ = false;
-    apollo::common::PointENU start_point;
-    if (!map_service_->GetStartPoint(&start_point)) {
-      AWARN << "Failed to get a dummy start point from map!";
-      return;
-    }
-    point.mutable_path_point()->set_x(start_point.x());
-    point.mutable_path_point()->set_y(start_point.y());
-    point.mutable_path_point()->set_z(start_point.z());
-    double theta = 0.0;
-    double s = 0.0;
-    map_service_->GetPoseWithRegardToLane(start_point.x(), start_point.y(),
-                                          &theta, &s);
-    point.mutable_path_point()->set_theta(theta);
-    point.set_v(start_velocity);
-    point.set_a(start_acceleration);
-  } else {
```

Comment:
Wondering why the else statement is removed? There are use cases of simulating control from the existing ADC location.

Reply:
I don't expect the synchronous version on initialization to work often. There is a use case where someone may want to switch from a recorded pose to a simulated pose. In a later commit, I added a callback for accepting localization.

Reply:
Agree, there is a use case of switching from a recorded pose to a simulated pose. From my understand, I see issues of handling this use case with the OnLocalization() in the latest commit:
1. While the component is able to get the latest recorded pose from OnLocalization(), the InitStartPoint() will ignore the SetStartPoint() from OnLocalization() and override it with the dummy point from map. Hence, the start point is always from dummy point from map.
2. This minor, but from performance perspective, we only need to get the data from localization when setting init start point. With the logic from the latest commit, we are reading the localization every 0.01 sec, which can be avoid.

Reply:
That is by design. Sorry, I didn't explain it well.

When this library was part of dreamview, it was getting localization messages in the background before sim_control was enabled. In that case, when it eventually called init there was a realistic duration where a localization message could be received: https://github.com/ApolloAuto/apollo/blob/10286356fc2c16be0b3d03e6ad6bd27220df95ce/modules/dreamview/backend/sim_control/sim_control.cc#L254. In the case where Init is called when the process just starts, there is very little time for localization messages to be received. Therefore, it is only handled asynchronously. This change also has the advantage of allowing for manually playing a bag to reset the position to a precise localization without toggling sim_control.